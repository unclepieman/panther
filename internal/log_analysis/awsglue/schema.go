package awsglue

/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

// Infers Glue table column types from Go types, recursively descends types

import (
	"github.com/panther-labs/panther/internal/log_analysis/awsglue/glueschema"
)

type Column = glueschema.Column

var (
	// RuleMatchColumns are columns added by the rules engine
	RuleMatchColumns = []Column{
		{
			Name:    "p_rule_id",
			Type:    glueschema.TypeString,
			Comment: "Rule id",
		},
		{
			Name:    "p_alert_id",
			Type:    glueschema.TypeString,
			Comment: "Alert id",
		},
		{
			Name:    "p_alert_context",
			Type:    glueschema.TypeString,
			Comment: "Additional alert context",
		},
		{
			Name:    "p_alert_creation_time",
			Type:    glueschema.TypeTimestamp,
			Comment: "The time the alert was initially created (first match)",
		},
		{
			Name:    "p_alert_update_time",
			Type:    glueschema.TypeTimestamp,
			Comment: "The time the alert last updated (last match)",
		},
		{
			Name:    "p_rule_tags",
			Type:    glueschema.ArrayOf(glueschema.TypeString),
			Comment: "The tags of the rule that generated this alert",
		},
		{
			Name:    "p_rule_reports",
			Type:    glueschema.MapOf(glueschema.TypeString, glueschema.ArrayOf(glueschema.TypeString)),
			Comment: "The reporting tags of the rule that generated this alert",
		},
	}

	// RuleErrorColumns are columns added by the rules engine
	RuleErrorColumns = append(
		RuleMatchColumns,
		Column{
			Name:    "p_rule_error",
			Type:    glueschema.TypeString,
			Comment: "The rule error",
		},
	)
)
