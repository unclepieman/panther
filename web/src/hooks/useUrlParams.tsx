/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import useRouter from 'Hooks/useRouter';
import queryString from 'query-string';
import omitBy from 'lodash/omitBy';

export const queryStringOptions = {
  arrayFormat: 'bracket' as const,
  parseNumbers: true,
  parseBooleans: true,
};

function useUrlParams<T extends { [key: string]: any }>() {
  const { history, location } = useRouter();

  /**
   * parses the query params of a URL and returns an object with params in the correct typo
   */
  const urlParams = queryString.parse(location.search, queryStringOptions) as T;

  /**
   * stringifies an object and adds it to the existing query params of a URL
   */
  const updateUrlParams = (params: Partial<T>) => {
    const mergedQueryParams = {
      ...urlParams,
      ...params,
    };

    // Remove any falsy value apart from the value `0` (number) and the value `false` (boolean)
    const cleanedMergedQueryParams = omitBy(
      mergedQueryParams,
      v => !v && !['number', 'boolean'].includes(typeof v)
    );

    history.replace({
      ...location,
      search: queryString.stringify(cleanedMergedQueryParams, queryStringOptions),
    });
  };

  /**
   * stringifies an object and replaces the query params of a URL
   */
  const setUrlParams = (params: Partial<T>) => {
    // Remove any falsy value apart from the value `0` (number) and the value `false` (boolean)
    const cleanedQueryParams = omitBy(params, v => !v && !['number', 'boolean'].includes(typeof v));

    history.replace({
      ...location,
      search: queryString.stringify(cleanedQueryParams, queryStringOptions),
    });
  };

  // Cache those values as long as URL parameters are the same
  return React.useMemo(
    () => ({
      urlParams,
      updateUrlParams,
      setUrlParams,
    }),
    [history.location.search]
  );
}

export default useUrlParams;
