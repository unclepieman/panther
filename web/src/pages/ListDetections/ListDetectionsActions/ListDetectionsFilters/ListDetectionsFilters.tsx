/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import urls from 'Source/urls';
import { Form, Formik, FastField } from 'formik';
import { SortDirEnum, ListDetectionsInput, ListDetectionsSortFieldsEnum } from 'Generated/schema';
import { Box, Flex } from 'pouncejs';
import pick from 'lodash/pick';
import useRequestParamsWithPagination from 'Hooks/useRequestParamsWithPagination';
import LinkIconButton from 'Components/buttons/LinkIconButton';
import FormikAutosave from 'Components/utils/Autosave';
import FormikCombobox from 'Components/fields/ComboBox';
import FormikTextInput from 'Components/fields/TextInput';
import DropdownFilters from './DropdownFilters';

export type ListDetectionsInlineFiltersValues = Pick<ListDetectionsInput, 'sortBy' | 'sortDir'>;

export type SortingOptions = {
  opt: string;
  resolution: ListDetectionsInput;
}[];

const filters = ['nameContains', 'sortBy', 'sortDir'] as (keyof ListDetectionsInput)[];

const defaultValues = {
  nameContains: '',
  sorting: null,
};

const sortingOpts: SortingOptions = [
  {
    opt: 'Display name (A-Z)',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.DisplayName,
      sortDir: SortDirEnum.Ascending,
    },
  },
  {
    opt: 'Display name (Z-A)',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.DisplayName,
      sortDir: SortDirEnum.Descending,
    },
  },
  {
    opt: 'Most Recently Modified',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.LastModified,
      sortDir: SortDirEnum.Descending,
    },
  },
  {
    opt: 'Oldest Modified',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.LastModified,
      sortDir: SortDirEnum.Ascending,
    },
  },
  {
    opt: 'Info to Critical',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.Severity,
      sortDir: SortDirEnum.Ascending,
    },
  },
  {
    opt: 'Critical to Info',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.Severity,
      sortDir: SortDirEnum.Descending,
    },
  },
  {
    opt: 'Enabled to Disabled',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.Enabled,
      sortDir: SortDirEnum.Ascending,
    },
  },
  {
    opt: 'Disabled to Enabled',
    resolution: {
      sortBy: ListDetectionsSortFieldsEnum.Enabled,
      sortDir: SortDirEnum.Descending,
    },
  },
];

const sortingItems = sortingOpts.map(sortingOption => sortingOption.opt);

/**
 * Since sorting is not responding to some ListDetectionsInput key we shall extract
 * this information from `sortBy` and `sortDir` parameters in order to align the
 * combobox values.
 */
const extractSortingOpts = params => {
  const { sorting, ...rest } = params;
  const sortingParams = sortingOpts.find(param => param.opt === sorting);
  return {
    ...rest,
    ...(sortingParams ? { ...sortingParams.resolution } : {}),
  };
};

const wrapSortingOptions = params => {
  const { sortBy, sortDir, ...rest } = params;
  const option = sortingOpts.find(
    param => param.resolution.sortBy === sortBy && param.resolution.sortDir === sortDir
  );

  return {
    ...(option ? { sorting: option.opt } : {}),
    ...rest,
  };
};

const ListDetectionsFilters: React.FC = () => {
  const { requestParams, updateRequestParamsAndResetPaging } = useRequestParamsWithPagination<
    ListDetectionsInput
  >();
  const initialFilterValues = React.useMemo(
    () =>
      ({
        ...defaultValues,
        ...wrapSortingOptions(pick(requestParams, filters)),
      } as ListDetectionsInlineFiltersValues),
    [requestParams]
  );
  return (
    <Flex justify="flex-end" align="center" spacing={4}>
      <Formik<ListDetectionsInlineFiltersValues>
        enableReinitialize
        initialValues={initialFilterValues}
        onSubmit={(values: ListDetectionsInlineFiltersValues) => {
          updateRequestParamsAndResetPaging(extractSortingOpts(values));
        }}
      >
        <Form>
          <FormikAutosave threshold={200} />
          <Flex spacing={4} align="center">
            <Box minWidth={425} maxWidth={490} flexGrow={3}>
              <FastField
                name="nameContains"
                icon="search"
                iconAlignment="left"
                as={FormikTextInput}
                label="Filter detections by text"
                placeholder="Search for a detection..."
              />
            </Box>
            <Box minWidth={225}>
              <FastField
                name="sorting"
                as={FormikCombobox}
                items={sortingItems}
                label="Sort By"
                placeholder="Select a sort option"
              />
            </Box>
          </Flex>
        </Form>
      </Formik>
      <DropdownFilters />
      <LinkIconButton
        icon="add-circle"
        aria-label="Create a new Detection"
        to={urls.detections.create()}
      />
    </Flex>
  );
};

export default React.memo(ListDetectionsFilters);
