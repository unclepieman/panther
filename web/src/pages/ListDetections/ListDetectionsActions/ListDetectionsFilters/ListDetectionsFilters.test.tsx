/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import queryString from 'query-string';
import { queryStringOptions } from 'Hooks/useUrlParams';
import { fireClickAndMouseEvents, fireEvent, render, within, waitFor } from 'test-utils';
import { ListDetectionsSortFieldsEnum, SortDirEnum } from 'Generated/schema';
import ListDetectionsFilters from './ListDetectionsFilters';

// Mock debounce so it just executes the callback instantly
jest.mock('lodash/debounce', () => jest.fn(fn => fn));

const parseParams = (search: string) => queryString.parse(search, queryStringOptions);

describe('ListDetectionsFilters', () => {
  const filtersUrlParams = `?analysisTypes[]=RULE&complianceStatus=ERROR&enabled=true&hasRemediation=true&initialSet=true&nameContains=AWS&page=1&sortBy=${ListDetectionsSortFieldsEnum.Severity}&sortDir=${SortDirEnum.Ascending}`;

  it('renders', () => {
    const { container, getByAriaLabel } = render(<ListDetectionsFilters />);
    expect(getByAriaLabel('Create a new Detection')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('initializes correctly from url params', () => {
    const { getByLabelText, getAllByLabelText, getByAriaLabel, getByTestId } = render(
      <ListDetectionsFilters />,
      {
        initialRoute: filtersUrlParams,
      }
    );
    expect(getByLabelText('Filter detections by text')).toHaveValue('AWS');
    expect(getAllByLabelText('Sort By')[0]).toHaveValue('Info to Critical');

    fireClickAndMouseEvents(getByAriaLabel('Detection Options'));

    const withinDropdown = within(getByTestId('dropdown-detections-listing-filters'));
    expect(withinDropdown.getAllByLabelText('State')[0]).toHaveValue('Enabled');
    expect(withinDropdown.getAllByLabelText('Policy Status')[0]).toHaveValue('Error');
    expect(withinDropdown.getAllByLabelText('Remediation Status')[0]).toHaveValue('Configured');
    expect(withinDropdown.getAllByLabelText('Created by')[0]).toHaveValue('Panther (system)');
    expect(withinDropdown.getByText('Rule')).toBeInTheDocument();
  });

  it('updates url params when filters change or get cleared', async () => {
    const {
      getAllByLabelText,
      getByText,
      history,
      getByLabelText,
      getByAriaLabel,
      getByTestId,
    } = render(<ListDetectionsFilters />);

    const dropdownTrigger = getByAriaLabel('Detection Options');

    fireEvent.change(getByLabelText('Filter detections by text'), { target: { value: 'AWS' } });

    fireClickAndMouseEvents(getAllByLabelText('Sort By')[0]);
    fireClickAndMouseEvents(getByText('Info to Critical'));

    fireClickAndMouseEvents(dropdownTrigger);

    // wait for debounce to kick in and update some stuff and then continue
    await waitFor(() => expect(history.location.search).toContain('nameContains'));

    const dropdown = getByTestId('dropdown-detections-listing-filters');
    let withinDropdown = within(dropdown);

    fireClickAndMouseEvents(withinDropdown.getAllByLabelText('State')[0]);
    fireClickAndMouseEvents(withinDropdown.getByText('Enabled'));

    fireClickAndMouseEvents(withinDropdown.getAllByLabelText('Policy Status')[0]);
    fireClickAndMouseEvents(withinDropdown.getByText('Error'));

    fireClickAndMouseEvents(withinDropdown.getAllByLabelText('Remediation Status')[0]);
    fireClickAndMouseEvents(withinDropdown.getByText('Configured'));

    fireClickAndMouseEvents(withinDropdown.getAllByLabelText('Created by')[0]);
    fireClickAndMouseEvents(withinDropdown.getByText('Panther (system)'));

    fireClickAndMouseEvents(withinDropdown.getAllByLabelText('Detection Types')[0]);
    fireClickAndMouseEvents(withinDropdown.getByText('Rule'));

    fireClickAndMouseEvents(withinDropdown.getByText('Apply Filters'));
    await waitFor(() =>
      expect(parseParams(history.location.search)).toEqual(parseParams(filtersUrlParams))
    );

    await waitFor(() => expect(dropdown).not.toBeInTheDocument());

    fireClickAndMouseEvents(dropdownTrigger);

    withinDropdown = within(getByTestId('dropdown-detections-listing-filters'));

    fireClickAndMouseEvents(withinDropdown.getByText('Clear Filters'));
    fireClickAndMouseEvents(withinDropdown.getByText('Apply Filters'));

    // expect to find the ONLY the filters that are not part of the dropdown
    await waitFor(() =>
      expect(parseParams(history.location.search)).toEqual({
        nameContains: 'AWS',
        page: 1,
        sortBy: 'severity',
        sortDir: 'ascending',
      })
    );
  });
});
