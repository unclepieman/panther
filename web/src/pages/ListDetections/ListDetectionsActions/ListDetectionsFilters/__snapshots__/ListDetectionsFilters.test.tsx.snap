// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDetectionsFilters renders 1`] = `
.panther-22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.panther-22>*:not(:last-child) {
  margin-right: 16px;
}

.panther-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.panther-16>*:not(:last-child) {
  margin-right: 16px;
}

.panther-6 {
  min-width: 425px;
  max-width: 490px;
  -webkit-box-flex: 3;
  -webkit-flex-grow: 3;
  -ms-flex-positive: 3;
  flex-grow: 3;
}



.panther-4 {
  min-height: 47px;
  position: relative;
  border: 1px solid;
  -webkit-transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  background-color: transparent;
  border-radius: 4px;
  border-color: #3e516b;
}

.panther-4:hover {
  border-color: #0081c5;
}

.panther-4:focus-within {
  border-color: #0081c5;
}

.panther-4:focus-within label {
  font-weight: 500;
  -webkit-transform: translate(6px, 4px) scale(0.65);
  -moz-transform: translate(6px, 4px) scale(0.65);
  -ms-transform: translate(6px, 4px) scale(0.65);
  transform: translate(6px, 4px) scale(0.65);
}

.panther-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.panther-0 {
  vertical-align: top;
  width: 100%;
  height: 100%;
  padding-left: 42px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: relative;
  color: #f6f6f6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  z-index: 2;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.panther-0::-webkit-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-0::-moz-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-0:-ms-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-0::placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}



.panther-0:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.panther-0:focus::-moz-placeholder {
  opacity: 0.4;
}

.panther-0:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.panther-0:focus::placeholder {
  opacity: 0.4;
}

.panther-0:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #f6f6f6;
  border-radius: 4px;
}

.panther-1 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #bdbdbd;
  top: 0;
  left: 26px;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.panther-2 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: currentColor;
  z-index: 1;
  position: absolute;
  left: 12px;
}

.panther-15 {
  min-width: 225px;
}

.panther-11 {
  position: relative;
}

.panther-7 {
  vertical-align: top;
  width: 100%;
  height: 100%;
  padding-left: 16px;
  padding-right: 32px;
  padding-top: 20px;
  padding-bottom: 8px;
  position: relative;
  color: #f6f6f6;
  font-size: 0.875rem;
  font-weight: 500;
  background-color: transparent;
  border: 0;
  cursor: pointer;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.panther-7::-webkit-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-7::-moz-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-7:-ms-input-placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}

.panther-7::placeholder {
  opacity: 0;
  color: #f6f6f6;
  -webkit-transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 500;
}



.panther-7:focus::-webkit-input-placeholder {
  opacity: 0.4;
}

.panther-7:focus::-moz-placeholder {
  opacity: 0.4;
}

.panther-7:focus:-ms-input-placeholder {
  opacity: 0.4;
}

.panther-7:focus::placeholder {
  opacity: 0.4;
}

.panther-7:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 30px #192231 inset;
  -webkit-text-fill-color: #f6f6f6;
  border-radius: 4px;
}

.panther-8 {
  pointer-events: none;
  font-size: 0.875rem;
  padding-left: 16px;
  padding-right: 16px;
  color: #bdbdbd;
  top: 0;
  left: 0;
  position: absolute;
  transform-origin: center left;
  -webkit-transform: translate(0, 14px) scale(1);
  -moz-transform: translate(0, 14px) scale(1);
  -ms-transform: translate(0, 14px) scale(1);
  transform: translate(0, 14px) scale(1);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: 400;
}

.panther-10 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  color: currentColor;
  opacity: 1;
  position: absolute;
  pointer-events: none;
  margin-top: auto;
  margin-bottom: auto;
  top: 0;
  bottom: 0;
  right: 16px;
}

.panther-18 {
  width: auto;
  outline: none;
  padding-left: 20px;
  padding-right: 20px;
  height: 46px;
  font-size: 0.938rem;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.panther-18:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.panther-18:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.panther-18:active,
.panther-18[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

.panther-17 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  color: currentColor;
  margin-right: -4px;
  margin-left: 12px;
}



.panther-21>span {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.panther-20 {
  outline: none;
  line-height: 1;
  border-radius: 4px;
  height: 46px;
  width: 46px;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.panther-20:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.panther-20:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.panther-20:active,
.panther-20[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

.panther-19 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  color: currentColor;
}

<div>
  <header
    id="main-header"
  />
  <div
    class="panther-22"
  >
    <form
      action="#"
    >
      <div
        class="panther-16"
      >
        <div
          class="panther-6"
        >
          <div
            class="panther-5"
          >
            <div
              class="panther-4"
            >
              <div
                class="panther-3"
              >
                <input
                  aria-invalid="false"
                  class="panther-0"
                  id="nameContains"
                  name="nameContains"
                  placeholder="Search for a detection..."
                  type="text"
                  value=""
                />
                <label
                  class="panther-1"
                  for="nameContains"
                >
                  Filter detections by text
                </label>
                <svg
                  class="panther-2"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14H14.71L14.43 13.73C15.41 12.59 16 11.11 16 9.5C16 5.91 13.09 3 9.5 3C5.91 3 3 5.91 3 9.5C3 13.09 5.91 16 9.5 16C11.11 16 12.59 15.41 13.73 14.43L14 14.71V15.5L19 20.49L20.49 19L15.5 14V14ZM9.5 14C7.01 14 5 11.99 5 9.5C5 7.01 7.01 5 9.5 5C11.99 5 14 7.01 14 9.5C14 11.99 11.99 14 9.5 14Z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="panther-15"
        >
          <div
            class="panther-5"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-0-label"
              class="panther-11"
              role="combobox"
            >
              <div
                class="panther-11"
              >
                <div
                  aria-disabled="false"
                  class="panther-4"
                >
                  <input
                    aria-autocomplete="list"
                    aria-disabled="false"
                    aria-invalid="false"
                    aria-labelledby="downshift-0-label"
                    aria-readonly="true"
                    autocomplete="off"
                    class="panther-7"
                    id="downshift-0-input"
                    name="sorting"
                    placeholder="Select a sort option"
                    readonly=""
                    type="text"
                    value=""
                  />
                  <label
                    class="panther-8"
                    for="downshift-0-input"
                    id="downshift-0-label"
                  >
                    Sort By
                  </label>
                </div>
                <svg
                  class="panther-10"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10L12 15L17 10H7Z"
                  />
                </svg>
              </div>
              <ul
                aria-labelledby="downshift-0-label"
                class="panther-5"
                id="downshift-0-menu"
                role="listbox"
              />
            </div>
          </div>
        </div>
      </div>
    </form>
    <button
      aria-describedby="1"
      aria-label="Detection Options"
      class="panther-18"
      type="button"
    >
      Filters 
      
      <svg
        class="panther-17"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M17.7865429,4 L6.2111062,4 C5.05130823,4 4.1111062,4.94020203 4.1111062,6.1 L4.11891921,6.28101292 C4.15525,6.70101378 4.31737448,7.10192992 4.58619539,7.43028751 L8.412,12.104 L8.41290323,18.8250834 C8.41290323,19.9848814 9.35310525,20.9250834 10.5129032,20.9250834 C10.8875811,20.9250834 11.255436,20.8248406 11.5783135,20.6347522 L14.4750877,18.9293279 L14.6185609,18.8368309 C15.1743482,18.4456922 15.5096774,17.8063253 15.5096774,17.1196592 L15.509,12.114 L19.4000354,7.44411387 C20.142369,6.5530066 20.0217641,5.22884114 19.1306568,4.48650755 C18.7532959,4.1721485 18.2776873,4 17.7865429,4 Z M17.7865429,6 C17.8099307,6 17.8325788,6.00819755 17.8505483,6.02316703 C17.8823736,6.04967894 17.8935604,6.09177564 17.8822518,6.1290687 L17.8633759,6.16400542 L13.5328444,11.3624339 L13.5156474,11.3924048 L13.5096774,11.4264393 L13.5096774,17.1196592 C13.5096774,17.1432747 13.5013484,17.1657153 13.486751,17.1833767 L13.4604112,17.2058339 L10.563637,18.9112581 C10.5482619,18.92031 10.530745,18.9250834 10.5129032,18.9250834 C10.4714819,18.9250834 10.4359425,18.8998995 10.4207617,18.864008 L10.4129032,18.8250834 L10.4129032,11.4259572 L10.4070777,11.3923244 L10.3902799,11.3626101 L6.1337295,6.16334702 C6.11909979,6.14547722 6.1111062,6.12309455 6.1111062,6.1 C6.1111062,6.05857864 6.13629019,6.02303932 6.17218169,6.0078585 L6.2111062,6 L17.7865429,6 Z"
        />
      </svg>
    </button>
    <a
      class="panther-21"
      href="/detections/new/"
    >
      <span
        aria-label="Create a new Detection"
        aria-pressed="false"
        class="panther-20"
        type="button"
      >
        <svg
          class="panther-19"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.92893219,4.92893219 C8.83417511,1.02368927 15.1658249,1.02368927 19.0710678,4.92893219 C22.9763107,8.83417511 22.9763107,15.1658249 19.0710678,19.0710678 C15.1658249,22.9763107 8.83417511,22.9763107 4.92893219,19.0710678 C1.02368927,15.1658249 1.02368927,8.83417511 4.92893219,4.92893219 Z M17.6568543,6.34314575 C14.5326599,3.21895142 9.46734008,3.21895142 6.34314575,6.34314575 C3.21895142,9.46734008 3.21895142,14.5326599 6.34314575,17.6568543 C9.46734008,20.7810486 14.5326599,20.7810486 17.6568543,17.6568543 C20.7810486,14.5326599 20.7810486,9.46734008 17.6568543,6.34314575 Z M13,7 L13,11 L17,11 L17,13 L13,13 L13,17 L11,17 L11,13 L7,13 L7,11 L11,11 L11,7 L13,7 Z"
          />
        </svg>
      </span>
    </a>
  </div>
  <footer
    id="footer"
  />
</div>
`;
